src/nodes.coffee

```diff
@@ -2209,7 +2209,9 @@ exports.Assign = class Assign extends Base
     answer = compiledName.concat @makeCode(" #{ @context or '=' } "), val
     # Per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assignment_without_declaration,
     # if we’re destructuring without declaring, the destructuring assignment must be wrapped in parentheses.
-    if o.level > LEVEL_LIST or o.level is LEVEL_TOP and isValue and @variable.base instanceof Obj and not @nestedLhs and not (@param is yes)
+    # The assignment is wrapped in parentheses if 'o.level' has lower precedence than LEVEL_LIST (3)
+    # (i.e. LEVEL_COND (4), LEVEL_OP (5) or LEVEL_ACCESS (6)), or if we're destructuring object, e.g. {a,b} = obj.
+    if o.level > LEVEL_LIST or isValue and @variable.base instanceof Obj and not @nestedLhs and not (@param is yes)
```

Comment:
This looks good to me. I'd suggest it could be a bit more readable if the condition was rearranged slightly:

```js
if o.level > LEVEL_LIST or (isValue and @variable.isObject() and not @nestedLhs and not @param)
```

Reply:
I appreciate the comment, but can you please explain why parentheses are required for those lower levels?

Reply:
@connec  `@variable.base instaceof Obj` is needed for cases like `{}.p = 0`. 
Changing `not (@param is yes)` to `not @param` cause two tests to fail, but I can't figure which ones exactly and why   ¯\\\_(ツ)\_/¯

@GeoffreyBooth I'm not sure exactly about lower levels. But, it seems that values for `LEVEL_COND` (e,g. `Existence`, `If`,...) and `LEVEL_ACCESS` (e.g. `Slice`, `In`) are already wrapped in parentheses elsewhere, before passing thru the `Assign`.
This leaves only `LEVEL_OP`, (e.g. `+`, `-`, ...) as in this case `a = b - c = 2` which has to be compiled to `a = b - (c = 2)`.
So, changing condition from `o.level > LEVEL_LIST` to `o.level is LEVEL_OP` works and all tests pass.   
On the other hand, wrapping expressions into another pair of parentheses doesn't hurt either.  
Perhaps someone else, more familiar with the internals, could shed some light on this.